### ISSUE TITLE
Unable to track video when I use drag and drop functionality and skip video directly from start to 25% or 50% or 75% milestone

### ISSUE THREAD

**Issue by SaurabhAloola** [yuhui/launchext-youtube-playback#34]
**Describe the bug**

**To Reproduce**
1. Go to 'site and drag video from start to 25%, 50% or 75% completion'
2. Check if the video milestone is tracked or not.

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [e.g. Windows]
 - Browser [e.g. Chrome, Safari]

**Smartphone (please complete the following information):**
 - Browser [e.g. stock browser, Safari]

**Additional context**

**Comment by yuhui** [comment]
@SaurabhAloola can you please provide the URL of a web page where I can replicate this problem?

**Comment by SaurabhAloola** [comment]
@yuhui Due to confidentiality, I can't share the URL with you. But I am having another issue with tracking. 

I have followed the below guide to implement the tracking. 
[](https://www.youtube.com/watch?v=wBpi7BoafQc)

However, the rules do not always trigger on the page. And it does occasionally fire on the page. When it triggers, everything works perfectly, and we get all of the desired rules with the appropriate evars and events.

We also checked with **Dom Ready** event with **"Enable video playback tracking"** action type. Issue still persists.
Can you please guide us if we are missing something here.

**Comment by SaurabhAloola** [comment]
Hi @yuhui 
Here I am Providing answer of the above questions.
2. I am providing screen recording for my Launch setup [Recording]([Edit Library | Tags | Adobe Experience Platform Data 
Collection - Awesome Screenshot](https://www.awesomescreenshot.com/video/27063062))
4. On my site videos are embedded in the web page code when the web page is loaded. You can check it on this site [site](https://www.insight.com/en_US/shop/partner/microsoft/hardware/hololens.html)
5. Yes, The same problem happen if I test in a private window.
6. Yes, The same problem happen if I test in any browser.

**Comment by yuhui** [comment]
Hi @SaurabhAloola , I think you're using another YouTube extension: "YouTube Player Embed" by 33sticks. That is not developed by me.

My extension is "YouTube Playback".
<img width="306" alt="image" src="https://github.com/yuhui/launchext-youtube-playback/assets/562000/80295323-aa3e-456c-b291-85b039dde1ea">

**Comment by SaurabhAloola** [comment]
Hi @yuhui, 
1. I think I have shared the uploaded version of the recording. Not the AwesomeScreenshot I am attaching the uploaded version again. 
recording link: 
https://github.com/yuhui/launchext-youtube-playback/assets/136058056/7adbb757-1a04-4a89-a1d1-47c00132d498

2. I have shared the live site with you to check the <script> tag of your Launch embed code. and how we have implemented the tracking on site. 
3. We have used the "YouTube Player Embed" extension earlier and as there was issue with milestones, we have enabled tracking with "YouTube Playback". Currently, it is deployed on the development version of the site. You have to use the DTM switch to change the environment to see that implementation. 
Attached the screenshot for your reference: 
![image](https://github.com/yuhui/launchext-youtube-playback/assets/136058056/95c1d312-cbc1-4145-88cc-c054ab9c0c97)

Please let me know if you need any help.

**Comment by yuhui** [comment]
Hi @SaurabhAloola , thanks for the guidance.

1. I am able to see your video now. Your setup looks correct, as far as I can tell.

3. Unfortunately, I am not able to switch environments with the DTM Switch extension. As you can see in the my screenshot of the extension, I need to browse _your_ Launch property's Environments page so that the extension can pick up the available environments first. Only after that will I be able to switch environments at your website with the extension.
<img width="173" alt="image" src="https://github.com/yuhui/launchext-youtube-playback/assets/562000/e8abd3cf-add5-4eca-86ce-396fd7a17417">

Can you share the <script> snippet of your Development environment?
<img width="1286" alt="image" src="https://github.com/yuhui/launchext-youtube-playback/assets/562000/e32e818c-d6be-45d9-a58f-98c1451eeb83">

Furthermore, I noticed that you are also running Google Tag Manager in your website. Do you have video tracking in there too?

**Comment by SaurabhAloola** [comment]
Hi @yuhui, 
Thank you for looking into that. We have provided you with the configuration recording earlier; you can refer that as well. We are for now providing you the recording how our tags are firing on the page. You have to ignore the tags that are firing for video start and other events that is configured with **Youtube Embed Player** Extension. 
The rules are listed below.

"YouTube Video Start"
"YouTube Video Paused"


https://github.com/yuhui/launchext-youtube-playback/assets/136058056/5dcc9b7b-0c8c-493a-bd5e-0c253b470588

Yeah, Its look like we also have separate tracking for this video through Google Tag manager with Google analytics. Is there anything that might affect the implementation of YouTube playback extension.

**Comment by SaurabhAloola** [comment]
Hi @yuhui , 

We will not be able to share the development script, though we could give access to our lower-environment site URLs, and you can check on that. Tell me about your **time zone details**, please, so we can build the library in that environment on that time and you can investigate the issue their.

Thanks!

**Comment by yuhui** [comment]
@SaurabhAloola , my time zone is UTC +8. Thanks.

**Comment by SaurabhAloola** [comment]
@yuhui Can you please specify when it will be a good time to allot the site URL for testing. According to your timezone.

**Comment by yuhui** [comment]
@SaurabhAloola , I will be free on my weekday evenings at 9pm-11:59pm UTC+8. Thanks.

**Comment by SaurabhAloola** [comment]
Greetings @yuhui I am providing you the URL for your testing: [https://projectqa.insight.com/en_US/shop/partner/fleetsharp.html](https://projectqa.insight.com/en_US/shop/partner/fleetsharp.html) 

Please update us with your finding here ASAP.

**Comment by yuhui** [comment]
Hi @SaurabhAloola , thanks for sharing the link. I was able to replicate your problem. Here are my findings:

You are using 2 YouTube extensions in your Launch property: my YouTube Playback extension, and the YouTube Player Embed extension. The combination of both caused errors to be thrown in my browser console due to errors with the YouTube IFrame API. When I removed all of the YouTube Player Embed extension code in your Launch script, then the errors were gone.

But even without the errors, I was only able to see some of my extension's events being successfully triggered. I found that there is a Google Tag Manager script in the web page. When I removed that GTM <script> snippet, then I was able to see my extension working properly; all of the Video Play, Pause, Milestone and Ended events were triggering properly.

I am aware that my extension doesn't work reliably when Google Tag Manager is being used and video tracking is enabled there too. I suspect that this is because of "race conditions" between GTM and my extension to use the YouTube IFrame API. I don't plan to solve this problem because my extension is intended to work solely with Launch as the _only_ tag manager in a web page.

So, I can offer these 2 suggestions:
1. Choose 1 YouTube-related extension only. Either remove the YouTube Player Embed extension from your Launch property, or remove mine. Whichever you choose, ensure that you finally only have 1 YouTube-related extension in your Launch property.
2. If you must use Google Tag Manager for video tracking, then I suggest that you do _all_ of your video tracking through GTM only. This will reduce the conflicts with the YouTube IFrame API. For example: setup GTM Custom HTML tags to send Launch's Direct Calls `_satellite.track()` with every GTM video trigger, and then setup corresponding Direct Call rules in your Launch property. **Note: I do not provide any support for Google Tag Manager implementation whatsoever.**

**Comment by SaurabhAloola** [comment]
Hey @yuhui, 
We have turned off all video tracking on our website by disabling the tracking that was done with the YouTube Player Embed extension and the video engagement tracking that was in the Google Tag Manager.
We have released the video tracking implementation using your extension. Even so, the rules do not always seem to work as intended.
Could you look into this and assist in determining what the precise problem is?

Thank you.

**Comment by yuhui** [comment]
@SaurabhAloola , I think you still have GTM's video tracking being enabled. When I inspected the HTML of the YouTube player, I found various "data-gtm-yt-inspected-*" attributes. These attributes are set by GTM when video is tracked there.
<img width="971" alt="image" src="https://github.com/yuhui/launchext-youtube-playback/assets/562000/ad21b7e0-7fe7-4242-9115-c6b9801ab477">
When I blocked the GTM script for GTM-KZFSXB, then the video tracking Rules worked all of the time.

**Comment by SaurabhAloola** [comment]
Hi @yuhui, If we exclude the GTM Script for GTM-KZFSXB, then it will also exclude the other tracking as well that is currently working through Google Tag Manager for page views, etc. I have turned off the **Video Engagement** from the Enhanced Measurement setting in Google Analytics that stopped video tracking tag being fired.

**Comment by SaurabhAloola** [comment]
Hi @yuhui, 
Thank you for your support. We have removed GTM tracking, and now it works for video start and pause. In case user does not drag/skip the video milestones being tracked as expected. 

But still, when we try to skip the video by dragging from 0 to 75%, in this case, there are no 25% or 50% milestones being tracked. 

You can check the below url :  https://projectqa.insight.com/en_US/shop/partner/fleetsharp.html[https://projectqa.insight.com/en_US/shop/partner/fleetsharp.html]

**Comment by yuhui** [comment]
Hi @SaurabhAloola , it is expected that dragging the video slider will **not** track milestones. The simple reason is that the user is searching for a particular location in the video timeline, so the video itself is not being played actually. Milestone tracking works only when the video is playing normally.

So in your example, since you had dragged the slider to 75%, so the 25% and 50% milestones are _not_ tracked. However, when you go back to 0 and let the video play normally all the way to the end, then the 25%, 50% and 75% milestones should be tracked.

**Comment by SaurabhAloola** [comment]
Hi @yuhui, Okay I got it But I have one query if we are not capturing the milestones while dragging the video in this scenario we are getting unspecified value in the Adobe Analytics report so can you suggest any better solution to see accurate data in the report cause we are also getting video end and pause count with Video milestone dimension.

**Comment by yuhui** [comment]
@SaurabhAloola , make sure you're using the Video milestone dimension with the "Video Milestone" event from my extension. This is stated clearly in the data element's configuration page.
